Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp

```diff
@@ -5675,21 +5661,20 @@ bool WebGLRenderingContextBase::validateArrayBufferType(const char* functionName
     return true;
 }
 
-bool WebGLRenderingContextBase::validateTexFuncData(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset)
+bool WebGLRenderingContextBase::validateTexFuncData(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset, size_t* totalBytesRead)
 {
     // All calling functions check isContextLostOrPending, so a duplicate check is not
     // needed here.
-    if (!pixels) {
+    if (!pixels && disposition != NullAllowed) {
         ASSERT(disposition != NullNotReachable);
-        if (disposition == NullAllowed)
-            return true;
         synthesizeGLError(GraphicsContextGL::INVALID_VALUE, functionName, "no pixels");
         return false;
     }
 
-    if (!validateSettableTexInternalFormat(functionName, format))
+    if (pixels && !validateSettableTexInternalFormat(functionName, format))
         return false;
-    if (!validateArrayBufferType(functionName, type, pixels ? std::optional<JSC::TypedArrayType>(pixels->getType()) : std::nullopt))
+
+    if (pixels && !validateArrayBufferType(functionName, type, pixels ? std::optional<JSC::TypedArrayType>(pixels->getType()) : std::nullopt))
```

Comment:
`if (pixels && validate(..., pixels ? a: b))` does not make sense on the account that the subexpression is redundant.

The army buffer type validation must happen regardless of whether `pixels` is passed or not. This is part of the bug about historical early return that you are fixing. In other words, `validateArrayBufferType` validates both `type` and the array buffer type.

Reply:
Currently we have sequence of
- validate parameters and data ptr, adjust them, come up with a new parameter
- adjust parameters even further
- clamp the parameters with the new parameter

you could prototype if we could change that to:
- validate parameters and return the correct parameters

so this would look something like:
`std::optional<GCGLSpan<void> WebGLRenderingContextBase::validateTexFuncData<typename SizeType>(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset)`

We would then remove the template parameter SizeType when all ANGLE texture functions would accept a consistent data size parameter.

Reply:
I've implemented this suggestion and was able to unify and delete some code. 👍